src/site/content/en/blog/porting-gphoto2-to-the-web/index.md

```diff
@@ -0,0 +1,352 @@
+EM_CACHE = $(EMSCRIPTEN)/cache
+
+# Sysroot is always under the `sysroot` subfolder.
+SYSROOT = $(EM_CACHE)/sysroot
+
+# …
+
+# For all dependencies I've used the same ./configure command with the
+# earlier defined SYSROOT path as the --prefix.
+deps/%/Makefile: deps/%/configure
+        cd $(@D) && ./configure --prefix=$(SYSROOT) # …
+```
+
+With such configuration, I only needed to run `make install` in each dependency, which installed it under the sysroot, and then the libraries found each other automatically.
+
+## Dealing with dynamic loading
+
+As mentioned above, libgphoto2 uses libtool to enumerate and dynamically load I/O port adapters and camera libraries. For example, the code for loading I/O libraries looks like this:
+
+```c
+lt_dlinit ();
+lt_dladdsearchdir (iolibs);
+result = lt_dlforeachfile (iolibs, foreach_func, list);
+lt_dlexit ();
+```
+
+There are a few problems with this approach on the web:
+
+- There is no standard support for dynamic linking of WebAssembly modules. Emscripten has its [custom implementation](https://emscripten.org/docs/compiling/Dynamic-Linking.html) that can simulate the [`dlopen()`](https://man7.org/linux/man-pages/man3/dlopen.3.html) API used by libtool, but it requires you to build "main'' and "side" modules with different flags, and, specifically for dlopen, also to [preload the side modules into the emulated filesystem](https://emscripten.org/docs/compiling/Dynamic-Linking.html#runtime-dynamic-linking-with-dlopen) during the application start-up. It can be difficult to integrate those flags and tweaks into an existing autoconf build system with lots of dynamic libraries.
+- Even if the `dlopen()` itself is implemented, there's no way to enumerate all dynamic libraries in a certain folder on the web, because most HTTP servers don't expose directory listings for security reasons.
+- Linking dynamic libraries on the command line instead of enumerating in runtime can also lead to problems, such as the [duplicate symbols issue](https://github.com/emscripten-core/emscripten/issues/11985), that are caused by differences between representation of shared libraries in Emscripten and on other platforms.
+
+It's possible to adapt the build system to those differences and hardcode the list of dynamic plugins somewhere during the build, but an even easier way to solve all those issues is to avoid dynamic linking to begin with.
+
+Turns out, libtool abstracts away various [dynamic linking methods](https://www.gnu.org/software/libtool/manual/html_node/Module-loaders-for-libltdl.html) on different platforms, and even supports writing custom loaders for others. One of the built-in methods it supports is called ["Dlpreopening"](https://www.gnu.org/software/libtool/manual/html_node/Dlpreopening.html):
+
+**“Libtool provides special support for dlopening libtool object and libtool library files, so that their symbols can be resolved even on platforms without any dlopen and dlsym functions.  
```

Comment:
Not sure how to add newlines in here. The standard Markdown method of adding two spaces at the end of the line doesn't seem to work in 11ty, judging by preview.

@jpmedley do you have ideas?

Reply:
Oh, perhaps when I was looking last the rendering wasn't updated to the latest version yet (the one with two spaces in the end of the line).